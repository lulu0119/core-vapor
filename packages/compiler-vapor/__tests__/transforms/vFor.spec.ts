import { makeCompile } from './_utils'
import {
  type ForIRNode,
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformText,
  transformVBind,
  transformVFor,
  transformVOn,
} from '../../src'
import { NodeTypes } from '@vue/compiler-dom'

const compileWithVFor = makeCompile({
  nodeTransforms: [
    transformVFor,
    transformText,
    transformElement,
    transformChildren,
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
  },
})

describe('compiler: v-for', () => {
  test('basic v-for', () => {
    const { code, ir, vaporHelpers, helpers } = compileWithVFor(
      `<div v-for="item of items" :key="item.id" @click="remove(item)">{{ item }}</div>`,
    )

    expect(code).matchSnapshot()
    expect(vaporHelpers).contains('createFor')
    expect(helpers.size).toBe(0)
    expect(ir.template).toEqual(['<div></div>'])
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.FOR,
        id: 0,
        source: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'items',
        },
        value: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'item',
        },
        key: undefined,
        index: undefined,
        render: {
          type: IRNodeTypes.BLOCK,
          dynamic: {
            children: [{ template: 0 }],
          },
        },
        keyProp: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'item.id',
        },
      },
    ])
    expect(ir.block.returns).toEqual([0])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 0 }],
    })
    expect(ir.block.effect).toEqual([])
    expect((ir.block.operation[0] as ForIRNode).render.effect).lengthOf(1)
  })

  test('multi effect', () => {
    const { code } = compileWithVFor(
      `<div v-for="(item, index) of items" :item="item" :index="index" />`,
    )
    expect(code).matchSnapshot()
  })

  test('w/o value', () => {
    const { code } = compileWithVFor(`<div v-for=" of items">item</div>`)
    expect(code).matchSnapshot()
  })

  test.todo('object de-structured value', () => {
    const { code } = compileWithVFor(
      '<span v-for="({ id, value }) in items">{{ id }}{{ value }}</span>',
    )
    expect(code).matchSnapshot()
  })
})
