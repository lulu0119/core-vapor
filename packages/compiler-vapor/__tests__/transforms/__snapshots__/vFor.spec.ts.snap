// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: v-for > basic v-for 1`] = `
"import { children as _children, on as _on, renderEffect as _renderEffect, setText as _setText, createFor as _createFor, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n1 = _createFor(() => (_ctx.items), (_block) => {
    const n2 = t0()
    const { 0: [n3],} = _children(n2)
    _on(n3, "click", () => $event => (_ctx.remove(_block.s[0])))
    _renderEffect(() => {
      const [item] = _block.s
      _setText(n3, item)
    })
    return n2
  })
  return [n1]
}"
`;

exports[`compiler: v-for > multi effect 1`] = `
"import { children as _children, renderEffect as _renderEffect, setDynamicProp as _setDynamicProp, createFor as _createFor, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n1 = _createFor(() => (_ctx.items), (_block) => {
    const n2 = t0()
    const { 0: [n3],} = _children(n2)
    _renderEffect(() => {
      const [item, index] = _block.s
      _setDynamicProp(n3, "item", item)
    })
    _renderEffect(() => {
      const [item, index] = _block.s
      _setDynamicProp(n3, "index", index)
    })
    return n2
  })
  return [n1]
}"
`;

exports[`compiler: v-for > w/o value 1`] = `
"import { createFor as _createFor, template as _template } from 'vue/vapor';
const t0 = _template("<div>item</div>")

export function render(_ctx) {
  const n1 = _createFor(() => (_ctx.items), (_block) => {
    const n2 = t0()
    return n2
  })
  return [n1]
}"
`;
